src/Microsoft.AspNetCore.Mvc.Razor/Internal/CSharpCompiler.cs

```diff
@@ -1,47 +1,146 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Hosting;
 using Microsoft.AspNetCore.Mvc.Razor.Compilation;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.Emit;
 using Microsoft.CodeAnalysis.Text;
-using Microsoft.Extensions.Options;
+using Microsoft.Extensions.DependencyModel;
+using DependencyContextCompilationOptions = Microsoft.Extensions.DependencyModel.CompilationOptions;
 
 namespace Microsoft.AspNetCore.Mvc.Razor.Internal
 {
     public class CSharpCompiler
     {
-        private readonly CSharpCompilationOptions _compilationOptions;
-        private readonly CSharpParseOptions _parseOptions;
         private readonly RazorReferenceManager _referenceManager;
         private readonly DebugInformationFormat _pdbFormat = SymbolsUtility.SupportsFullPdbGeneration() ?
             DebugInformationFormat.Pdb :
             DebugInformationFormat.PortablePdb;
 
-        public CSharpCompiler(RazorReferenceManager manager, IOptions<RazorViewEngineOptions> optionsAccessor)
+        public CSharpCompiler(RazorReferenceManager manager, IHostingEnvironment hostingEnvironment)
+            : this(manager, hostingEnvironment, GetDependencyContextCompilationOptions(hostingEnvironment))
         {
-            _referenceManager = manager;
-            _compilationOptions = optionsAccessor.Value.CompilationOptions;
-            _parseOptions = optionsAccessor.Value.ParseOptions;
+        }
+
+        internal CSharpCompiler(
+            RazorReferenceManager manager,
+            IHostingEnvironment hostingEnvironment,
+            DependencyContextCompilationOptions dependencyContextOptions)
+        {
+            _referenceManager = manager ?? throw new ArgumentNullException(nameof(manager));
+            if (hostingEnvironment == null)
+            {
+                throw new ArgumentNullException(nameof(hostingEnvironment));
+            }
+
+            ParseOptions = GetParseOptions(hostingEnvironment, dependencyContextOptions);
+            CSharpCompilationOptions = GetCompilationOptions(hostingEnvironment, dependencyContextOptions);
             EmitOptions = new EmitOptions(debugInformationFormat: _pdbFormat);
         }
 
+        public CSharpParseOptions ParseOptions { get; }
```

Comment:
I don't think the callback allows you to modify the parse options (since it's working on the already parsed syntax tree), you should be able to affect the compilation options though.